extensible device registration docs (#401)
SoftwareAG/c8y-docs: content/concepts/applications-bundle/extensible-device-registration.md

---
weight: 50
title: Extensible device registration
layout: redirect
---
### Introduction

> To address the growing number of IoT protocols and certain restrictions in the general single or bulk device registration, an extensible device registration is available with release 10.15,  which enables you to easily add own device protocols to the device registration process in {{< product-c8y-iot >}}.

Reviewer: This sentence is a bit long.

Reviewer 2: removed the last part after `,`

The general concept is based on extending the device registration using a metadata-based approach. Microservices and agents that implement current device registrations can add custom forms to the device registration wizard by providing simple descriptions of the required registration attributes. The metadata is then used by the UI to render a corresponding device registration wizard.

There are two possible ways to extend the UI:
- with a [single device registration](/concepts/applications/#single-device-registration) form
- with a [bulk device registration](/concepts/applications/#bulk-device-registration) form

{{< c8y-admon-req >}}

> Extensible device registration requires [application extension](/concepts/applications/#extension-enabling) to be defined and the microservice to implement the predefined endpoints used for getting device registration metadata and creating the device.

Reviewer 3: ```suggestion
Extensible device registration requires the [application extension](/concepts/applications/#extension-enabling) to be defined, and the microservice to implement the predefined endpoints used for getting device registration metadata and creating the device.
```

{{< /c8y-admon-req >}}

### Advantages of extended device registration

The extended device registration provides the following advantages:

> - **Extensibility of the device registration wizard**

Reviewer: Why do you use enumerations here with html breaks? I'd simply use ":" without a line break and append this to the bullet. Or use proper subheaders (####)

Reviewer 2: @BeateRixen you did the formatting. What do you think?

Reviewer 3: > @BeateRixen you did the formatting. What do you think?

It was formatted as headings but that looked a bit "oversized". If you prefer, we can remove the line break and add a colon. We do not have too strict rules on such lists.

- **No UI code changes required**
<br>You do not need to write UI Angular code. This keeps the amount of integration work as little as possible. The device integration developer only needs to subscribe a microservice that provides an own wizard, and the wizard shows up automatically.

### Extension enabling

As a first step to extend the device registration flow you must define extensions in the application representation.

{{< c8y-admon-info >}}
The `extensions` fragment can be placed either as root level or inside the `manifest` fragment of the application representation.
{{< /c8y-admon-info >}}

There are two types of extensions:
 - Single device registrations type: `extensibleDeviceRegistration`, for example:

```json
"extensions": [
  {
    "name": "<extension name>",
    "description": "<description>",
    "type": "extensibleDeviceRegistration"
  },
  ...
]
```
- Bulk device registration type: `extensibleBulkDeviceRegistration`, for example:
```json
"extensions": [
  {
    "name": "<extension name>",
    "description": "<description>",
    "type": "extensibleBulkDeviceRegistration"
  },
  ...
]
```

### Single device registration

> After enabling the `extensibleDeviceRegistration` extension type, the Device management > Devices > Register device menu is being extended with an entry corresponding to the extension `name` property:

Reviewer 3: ```suggestion
After enabling the `extensibleDeviceRegistration` extension type, the **Devices** > **Register device** menu in the Device Management application is being extended with an entry corresponding to the extension `name` property:
```

                      "enum": [
                        "NO_SEC",
                        "SEC"
                      ]
                    },
                    "imei": {
                        "examples": [
                            "012345678901234"
                        ],
                        "type": "string",
                        "title": "Imei number"
                    }
                },
                "required": [
                    "imei",
                    "security"
                ]
            },
            {...}
        ]
    }
}
```
The important part is the `pages` array which contains steps of the wizard that the modal is going to render accordingly to the JSON Schema definition: [https://json-schema.org/](https://json-schema.org/).

As a result the following wizard will be displayed:

![Select guide](/images//concepts-guide/extensible-device-registration/extensible-single-device-reg.png)

In the final step all data collected via the wizard will be sent back to the microservice using the following REST endpoint:

`POST /service/<contextPath>/deviceRegistration`

```json
{
  "imei": "012345678901234",
  "security": "NO_SEC",
  ...
}
```

> The form is able to send anything defined via JSON Schema and the microservice which provides the form definition is responsible for the proper handling of the submitted data.

Reviewer: This sentence is somehow broken. Do you want to say the following?

> The form is able to send anything defined via JSON Schema **TO** the microservice which provides the form definition is responsible for the proper handling of the submitted data.

I'd probably also break this sentence up into two smaller ones for the sake of clarity.

Reviewer 2: done

#### API specification

The device integration microservices must implement the following REST endpoints:

```
GET /service/<contextPath>/deviceRegistration/metadata?lang=<user-language>
Accept: application/json
```
Returns the metadata in the vocabulary of the JSON Schema.

```
POST /service/<contextPath>/deviceRegistration
Content-type: application/json
```
Creates a single device based on the collected data. Sends application/json with key-value pairs.

#### Single device registration flow diagram

The following diagram visualizes the single device registration flow:

![Single diagram](/images//concepts-guide/extensible-device-registration/single-diagram.png)

### Bulk device registration

> The key functionality required for many device integrations is the ability to register many devices at the same time. Currently, all protocols have to rely on the bulk-registration mechanism of the platform, which often either requires too many fields or requires custom fields to be added. The latter ones can however so far not be validated, as the core directly creates devices -- and microservices and agents have no control over the properties being written to the managed objects.

Reviewer: > The key functionality required for many device integrations is the ability to register many devices at the same time. 

Too many words you don't need - this can be more concise:
>Many device integrations require the registration of many devices at the same time.

Reviewer 2: done

After enabling the `extensibleBulkDeviceRegistration` extension type, the Device management > Devices > Register device  `Bulk device registration` modal is being displayed with an extended wizard entry corresponding to the extension `name` property:

![Select guide](/images/concepts-guide/extensible-device-registration/register-device-menu-with-extensible-device-reg.png)

Additionally, the microservice provides the title of the wizard step and example bulk file(s):
```json
{
  "c8y_DeviceRegistration": {
    "title": "<title>",
    "description": "<description>",
    "bulk": {
      "title": "<bulk form title>",
      "exampleFileUrls": [{
        "title": "<example title>",
        "description": "<example description>",
        "url": "<publicly-reachable-URL>"
      }]
    }
  }
}
```

As a result the following wizard will be displayed:

![Select guide](/images/concepts-guide/extensible-device-registration/extensible-bulk-device-reg.png)

#### API specification

> The device integration microservices have to implement the following REST endpoints:

Reviewer 3: ```suggestion
The device integration microservices must implement the following REST endpoints:
```

```
GET /service/<contextPath>/deviceRegistration/metadata?lang=<user-language>
Accept: application/json
```

Returns the metadata in the vocabulary of the JSON Schema.

```
POST /service/<contextPath>/deviceRegistration/bulk`
Content-Type: multipart/form-data; boundary=boundary

--boundary
Content-Disposition: form-data; name="file"; filename="<input csv file>"
Content-Type: text/csv

--boundary--
```

Sends multipart form-data of the csv file type.

#### Bulk flow diagram

The following diagram visualizes the bulk device registration flow:

![Bulk diagram](/images//concepts-guide/extensible-device-registration/bulk-diagram-sync.png)

### Limitations

-   The concept does not allow the microservice to hook into deregistrations/decommissioning of a device:
    - Any device integration microservice has to check if a device was deleted, for example to perform garbage collection.

> - Interactive steps (like “Bluetooth Coupling”) can not be implemented as of now: The reason here is that the concept assumes only static properties are required for a device registration, not a sort of process that requires user interactions

Reviewer 3: ```suggestion
- Interactive steps (like “Bluetooth Coupling”) can not be implemented as of now. The reason here is that the concept assumes that only static properties are required for a device registration, not some process that requires user interactions.
```

>     - In theory, the meta-data based approach can easily be extended in a way that the JSON used to render the form is not static, but dynamically generated by the microservice. For example, you could specify a “page” that draws the next specification dynamically from an endpoint after posting all given user inputs there

Reviewer 3: ```suggestion
    - In theory, the meta-data based approach can easily be extended in a way that the JSON used to render the form is not static, but dynamically generated by the microservice. For example, you could specify a “page” that draws the next specification dynamically from an endpoint after posting all given user inputs there.
```

>     - We step back from such a solution from now, because it complicates the data model and the requirements to be fulfilled by the microservice.

Reviewer 3: ```suggestion
    - We step back from such a solution for now, because it complicates the data model and the requirements to be fulfilled by the microservice.
```

---
weight: 50
title: Extensible device registration
layout: redirect
---
### Introduction

Due to growing number of IoT protocols, Extensible Device Registration enables possibility 
to easily add own device protocol to device registration process.

The over all concept is to extend the device registration using a metadata based approach. Microservices and agents that implement current device registrations can add custom forms to the device registration wizard by providing a simple description of the required registration attributes. 

The metadata will be used by the UI to render a corresponding device registration wizard. There are two possible ways to extend the UI:
- with [single](#single) device registration form
- and [bulk](#bulk) device registration] form

{{< c8y-admon-info >}}

> As a precondition, application extension [{Configuration}}] definition have to defined and, in case of Single Device Registration, microservice realizing the registration have to provide additional metadata.

Reviewer 4: link for [{Configuration}}] is missing and you can remove ", in case of Single Device Registration," because it required metadata being present for both as far as I experienced.

---
weight: 50
title: Extensible device registration
layout: redirect
---
### Introduction

> To address the growing number of IoT protocols and certain restrictions in the general single or bulk device registration, an extensible device registration has been introduced which enables you to easily add own device protocols to the device registration process in {{< product-c8y-iot >}}.

Reviewer 3: ```suggestion
To address the growing number of IoT protocols and certain restrictions in the general single or bulk device registration, an extensible device registration is available with release 10.15,  which enables you to easily add own device protocols to the device registration process in {{< product-c8y-iot >}}.
```

The general concept is based on extending the device registration using a metadata-based approach. Microservices and agents that implement current device registrations can add custom forms to the device registration wizard by providing simple descriptions of the required registration attributes. The metadata is then used by the UI to render a corresponding device registration wizard.

There are two possible ways to extend the UI:
- with a [single device registration](/concepts/applications/#single-device-registration) form
- with a [bulk device registration](/concepts/applications/#bulk-device-registration) form

{{< c8y-admon-req >}}
As a precondition, the application extension [{Configuration}}] definition has to be defined and, in case of single device registration, the microservice realizing the registration has to provide additional metadata.
{{< /c8y-admon-required >}}

> ### The extension drivers

Reviewer 3: I don´t think that this section is relevant for the user. I mentioned in the introduction that there were "certain restrictions in the general single or bulk device registration". I think this is sufficient.

Reviewer 2: removed

### Extension enabling

As a first step to extend the device registration flow you must define extensions in the application representation.

{{< c8y-admon-info >}}
The `extensions` fragment can be placed either as root level or inside the `manifest` fragment of the application representation.
{{< /c8y-admon-info >}}

There are two types of extensions:
 - Single device registrations type: `extensibleDeviceRegistration`, for example:

```json
"extensions": [
  {
    "name": "<extension name>",
    "description": "<description>",
    "type": "extensibleDeviceRegistration"
  },
  ...
]
```
- Bulk device registration type: `extensibleBulkDeviceRegistration`, for example:
```json
"extensions": [
  {
    "name": "<extension name>",
    "description": "<description>",
    "type": "extensibleBulkDeviceRegistration"
  },
  ...
]
```

### Single device registration

After enabling the `extensibleDeviceRegistration` extension type, the Device management > Devices > Register device menu is being extended with an entry corresponding to the extension `name` property:
![Select guide](/images/extensible-device-registration/register-device-menu-with-extensible-device-reg.png)

> From now on, everything will be rendered based on data provided via the custom microservice. The added menu entry opens a window which fetches the form definition using the following endpoint:

Reviewer 3: It seems that something is missing here

Reviewer 2: Copied the endpoint below this text

> The context of the microservice, which is taken from application definition:

Reviewer 3: Also, this sentence sounds incomplete

Reviewer 2: rephrased

        "pages": [
            {
                "$schema": "https://json-schema.org/draft/2020-12/schema",
                "type": "object",
                "title": "Mandatory settings",
                "properties": {
                    "security": {
                      "default": "NO_SEC",
                      "type": "string",
                      "title": "Security mode",
                      "enum": [
                        "NO_SEC",
                        "SEC"
                      ]
                    },
                    "imei": {
                        "examples": [
                            "012345678901234"
                        ],
                        "type": "string",
                        "title": "Imei number"
                    }
                },
                "required": [
                    "imei",
                    "security"
                ]
            },
            {...}
        ]
    }
}
```
The important part is the `pages` array which contains steps of the wizard that the modal is going to render accordingly to the JSON Schema definition: [https://json-schema.org/](https://json-schema.org/).

> JSON Schema is a generic-purpose language to specify data, along with constraints and data types and simple validations.

Reviewer 3: I would either remove this sentence or add it when the JSON Schema is first mentioned.

Reviewer 2: removed

  "c8y_DeviceRegistration": {
    "title": "<title>",
    "description": "<description>",
    "bulk": {
      "title": "<bulk form title>",
      "exampleFileUrls": [{
        "title": "<example title>",
        "description": "<example description>",
        "url": "<publicly-reachable-URL>"
      }]
    }
  }
}
```

As a result the following wizard will be displayed:
![Select guide](/images/extensible-device-registration/extensible-bulk-device-reg.png)

#### API specification

The device integration microservices have to implement the following REST endpoints:

```
GET /service/<contextPath>/deviceRegistration/metadata?lang=<user-language>
Accept: application/json
```

Returns the metadata in the vocabulary of the JSON Schema.

```
POST /service/<contextPath>/deviceRegistration/bulk`
Content-Type: multipart/form-data; boundary=boundary

--boundary
Content-Disposition: form-data; name="file"; filename="<input csv file>"
Content-Type: text/csv

--boundary--
```

Sends multipart form-data of the csv file type.

#### Bulk flow diagram

The following diagram visualizes the bulk device registration flow:

![Bulk diagram](/images/extensible-device-registration/bulk-diagram-sync.png)

> ### Limitations

Reviewer 3: Is this whole section entirely relevant for the user?

Reviewer 2: yes.

```
POST /service/<contextPath>/deviceRegistration
Content-type: application/json
```
Creates a single device based on the collected data. Sends application/json with key-value pairs.

#### Single device registration flow diagram

The following diagram visualizes the single device registration flow:

![Single diagram](/images//concepts-guide/extensible-device-registration/single-diagram.png)

### Bulk device registration

The key functionality required for many device integrations is the ability to register many devices at the same time. Currently, all protocols have to rely on the bulk-registration mechanism of the platform, which often either requires too many fields or requires custom fields to be added. The latter ones can however so far not be validated, as the core directly creates devices -- and microservices and agents have no control over the properties being written to the managed objects.

After enabling the `extensibleBulkDeviceRegistration` extension type, the Device management > Devices > Register device  `Bulk device registration` modal is being displayed with an extended wizard entry corresponding to the extension `name` property:

> ![Select guide](/images//concepts-guide/extensible-device-registration/register-device-menu-with-extensible-device-reg.png)

Reviewer 3: No, please remove one /

> ![Select guide](/images//concepts-guide/extensible-device-registration/register-device-menu-with-extensible-device-reg.png)

Reviewer 2: done
